plugins/audio-s5gen2/fu-audio-s5gen2-device.c

```diff
@@ -47,88 +48,96 @@ fu_qc_s5gen2_device_msg_out(FuQcS5gen2Device *self, guint8 *data, gsize data_len
 }
 
 static gboolean
-fu_qc_s5gen2_device_msg_in(FuQcS5gen2Device *self, guint8 *data_in, gsize data_len, GError **error)
+fu_qc_s5gen2_device_msg_in(FuQcS5gen2Device *self,
+			   guint8 *data_in,
+			   gsize data_len,
+			   gsize *read_len,
+			   GError **error)
 {
 	FuDevice *proxy = fu_device_get_proxy(FU_DEVICE(self));
+	g_autoptr(GByteArray) err_msg = NULL;
+	g_autoptr(GError) error_local = NULL;
+
 	if (proxy == NULL) {
 		g_set_error_literal(error, FWUPD_ERROR, FWUPD_ERROR_NOT_SUPPORTED, "no proxy");
 		return FALSE;
 	}
-	return fu_qc_s5gen2_impl_msg_in(FU_QC_S5GEN2_IMPL(proxy), data_in, data_len, error);
+	if (!fu_qc_s5gen2_impl_msg_in(FU_QC_S5GEN2_IMPL(proxy), data_in, data_len, read_len, error))
+		return FALSE;
+
+	err_msg = fu_struct_qc_error_ind_parse(data_in, *read_len, 0, &error_local);
+	if (err_msg != NULL) {
+		guint16 code = fu_struct_qc_error_ind_get_error_code(err_msg);
+		/* error detected */
+		g_autoptr(GByteArray) confirm = fu_struct_qc_error_res_new();
+		fu_struct_qc_error_res_set_error_code(confirm, code);
+
+		/* confirm and stop */
+		if (!fu_qc_s5gen2_impl_msg_out(FU_QC_S5GEN2_IMPL(proxy),
+					       confirm->data,
+					       confirm->len,
+					       error))
+			return FALSE;
+
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_INVALID_DATA,
+			    "unexpected error (0x%.02X)",
```

Comment:
maybe `invalid code: 0x%x`

Reply:
mmm... no. Not just an invalid code.
Here I check every message with `fu_struct_qc_error_ind_parse()` above to recognize if we got the error message instead of any other reply -- meaning the device has some internal (and unexpected) error state, or we sent a wrong message at a wrong time. So it is literally an error indication, which shouldn't happens at all.

Reply:
okay, just 0x%x then please.